### ISSUE TITLE
Randomize the enabled features string in analytics

### ISSUE THREAD

**Issue by eliasyishak** [flutter/flutter#147327]
Currently, the enabled features we send for analytics remains in the same order that they were enabled. So if we had 3 features:
- my-first-feature
- my-second-feature
- my-thrid-feature

The list will always be = `"my-first-feature,my-second-feature,my-thrid-feature"`

In GA4 we have a 36 character limit so if the string is longer than 36 characters, anything after 36 gets truncated and dropped. This could send a false signal that a given feature is not being used,

We should instead randomize this list with each invocation so that each event has the chance to be sent.

**Comment by christopherfujino** [comment]
can we instead encode the features in a more compact way?

**Comment by eliasyishak** [comment]
Is there a way we can guarantee they won't exceed 36 characters? If so, that would be a good alternative

**Comment by christopherfujino** [comment]
> Is there a way we can guarantee they won't exceed 36 characters? If so, that would be a good alternative

I mean, fundamentally, there's only so much data you can encode in 36 characters, so eventually we might exceed this. However, I'm guessing we shouldn't use arbitrarily long strings to identify features if we're capped at 36 characters.

**Comment by loic-sharma** [comment]
> Is there a way we can guarantee they won't exceed 36 characters? If so, that would be a good alternative

Could we use an array of strings using [BigQuery's "repeated record" type](https://cloud.google.com/bigquery/docs/nested-repeated)?

> I'm guessing we shouldn't use arbitrarily long strings to identify features if we're capped at 36 characters.

The Swift Package Manager config, `enable-swift-package-manager` (28 characters), causes this truncation very quickly.

**Comment by christopherfujino** [comment]
> Could we use an array of strings using [BigQuery's "repeated record" type](https://cloud.google.com/bigquery/docs/nested-repeated)?

My (vague) understanding is that the cap of 36 characters is imposed by the GA API, before the data ever hits bigquery. Once it gets exported to BigQuery, I believe it does become some kind of repeated field. Would be happy to jump on a GVC with you to verify how the data is being plumbed.

**Comment by loic-sharma** [comment]
@zanderso suggested:

> Off the top of my head, my first thought would be to pull the feature flags out of a single user property, and instead send one event per (enabled?) flag.

**Comment by zanderso** [comment]
No, to clarify, I mean that each config flag would trigger sending its own event, and the flag name would go in the 'label' property of the event. So, like:

* `event('flutter-enabled-feature', label: 'enable-foo')`
* `event('flutter-enabled-feature', label: 'enable-bar')`
* ...

That way, you get the full 100 chars of the event property for each flag since they're each in a separate event. The downside is that the only way to correlate flags with each other is by way of the user id or session id.

**Comment by zanderso** [comment]
> Do you have a specific use case for this, or is it hypothetical?

It's hypothetical.

**Comment by bkonyi** [comment]
Dropping the priority to P2. We don't really have the resources to dig into this right now and haven't received much signal that it's critical.

**Comment by bkonyi** [comment]
After discussing with @loic-sharma, I'm raising this back to P1 as it's important for the iOS team. However, we still don't have the resources to dig into this and likely won't make any progress on it in the near future.

**Comment by andrewkolos** [comment]
**TL;DR**: We should send `enabled_features` as a event parameter (on all tool events) rather than as a user property. This gives us a length limit of 100 (500 if we can get our GA4 instance ("property") upgraded to GA4 360). This will be an immediate improvement to the existing situation, and it should hopefully be enough (indefinitely and for almost all users) if we keep feature names short.

Tasks for me:

- [ ] Verify whether on-by-default features like `enable-web`, `enable-linux-desktop`, `enable-android`, etc. are meant to be included in `enabled_features`. From poking around in data today, sometimes I see them, sometimes I don't. I feel like there is a bug somewhere. _If_ these should always be present, than the 100 length limit becomes less viable.
- [ ] Patch unified_analytics tool to include `enabled_features` on all tool events as an event parameter rather than a user property.
- [ ] Add a lint/test to the tool that makes sure that names for new features are limited in length (with exceptions for ones that exist today).
- [ ] See if we can get our GA4 Property upgraded to a 360 property to up the event parameter limit from 100 to 500.
______________

Re-summarizing the issue in my own understanding:

In every analytics event, a string-to-string map is included—called `user_properties`. The length of a user property value is 36. We can quickly run into this limit if a user has more than one or two features toggled.

A few possible solutions (some not mutually exclusive from each other):
1. Come up with an encoding scheme. Features could be abbreviated (e.g. `enable-native-assets` becomes `ena`)[^1].
2. Send a new, separate `enabled_feature` event *per* enabled feature *per* tool invocation (per  [@zanderso's earlier comment](https://github.com/flutter/flutter/issues/147327#issuecomment-2189989170)). This event would contain a single event parameter that indicates a feature was on. This would require all tool events include a unique ID, so that features could be correlated back to individual tool invocations. Note that `session_id` user property we send is not sufficient, since it is not unique to a CLI command invocation[^2]. While every GA4 event includes an `event_timestamp`, it represents the time that the _batch_ this event was a part of was sent[^3].
3. Send `enabled_features` as a event parameter (on all tool events) rather than as a user property. This gives us a length limit of 100 (500 if you get our GA4 Property upgraded to a 360 property).  Additionally, we can add a lint to make sure all new feature names are shorter than *N* characters to make this limit harder to bump into.

credit to @matanlurey for letting me bounce ideas off of him 👍 

[^1]: Purely for the sake of being exhaustive, another idea is to use bit-magic compress a near-infinite amount of binary data into 36 unicode characters. But the complexity on the querying-end alone would be nightmarish.

[^2]: https://github.com/dart-lang/tools/blob/7a231e576c9a2c057cbce035078d9a3dae6916c9/pkgs/unified_analytics/lib/src/user_property.dart#L67. 

[^3]: Docs: https://developers.google.com/analytics/devguides/collection/protocol/ga4/sending-events?client_type=firebase#override_timestamp

**Comment by zanderso** [comment]
> This would require all tool events include a unique ID, so that features could be correlated back to individual tool invocations.

To clarify a bit, what I was trying to say above is that I don't think this is necessary. Thinking ahead to what we want to learn from this data, it's sufficient to know the proportion of users with a feature enabled, and likely not necessary to know the exact combination of features enabled in each command.

**Comment by andrewkolos** [comment]
On second thought, without knowing of any concrete query use case where sending this information as separate event(s) is a pain, I think the separate event(s) approach is much simpler. I'll need to sleep on this.

**Comment by loic-sharma** [comment]
@andrewkolos Here are the questions we'd like to answer for Swift Package Manager:

1. Does the Swift Package Manager feature increase error rates?
2. Can we remove CocoaPods support from Flutter's tooling?
3. Can we tell plugin authors that they can remove CocoaPods integration from their plugins? (This is separate from the previous question as app developers might not be on the latest version of Flutter)

To answer these questions, here is the data we might collect on app builds:

1. Is SwiftPM on?
2. Does the app have SwiftPM integration? (Old projects need to be updated to support SPM)
3. How many plugins used CocoaPods?
4. How many plugins used Swift Package Manager?

**Comment by bkonyi** [comment]
Dropping priority to P2 as this isn't urgent and I don't have cycles to pick this up.

**Comment by CarlosMendonca** [comment]
Discussed this with @loic-sharma and we agree that solution 2 (from https://github.com/flutter/flutter/issues/147327#issuecomment-2492550071) is the best path forward to consistently capture enable flag telemetry. We should implement this soon, as there are several features still gated by their flags that we need to better understand usage of. @bkonyi, do you have cycles to work on this now and can we raise this back to P1?
